content/blog/2025-05-13-zihan-ethan-project.md

```diff
@@ -0,0 +1,133 @@
+
+
+#### Bril Fuzzer
+Rather than generating code at the Bril IR level, our fuzzer works on AST level with if-else and loop constructs. This lets us generate “interesting” bril programs with reducible CFGs and configurable nesting levels. Although the reducibility of cfg is not a requirement for dataflow analysis, we hope to fuzz IRs that resemble those emitted from real programs.
+
+For the same reason, we also limit the maximum nesting depth of basic blocks. In practice, most of the real-world programs won’t have loops that go over three levels deep. By enforcing this, we also limit the number of back edges within each SCC and the average component size. 
+
+(we recommend [bril2json-rs](https://github.com/sampsyo/bril/tree/main/bril2json-rs) for serializing large fuzzed bril programs, the default [python impl](https://github.com/sampsyo/bril/tree/main/bril-txt) for that is sometimes too slow)
+
+Our [bril fuzzer](https://github.com/zihan0822/para-dflow/tree/main/bril-fuzzer), [flattened bril repr](https://github.com/zihan0822/para-dflow/tree/main/bril) and the [parallel solver](https://github.com/zihan0822/para-dflow/tree/main/bril-analysis) are all open sourced on [Github](https://github.com/zihan0822/para-dflow/tree/main)
+
+## Parallel Dataflow Solver
+There are two main phases for our parallel solver:
+##### 1. Compute KILL and GEN set in parallel
+Besides flattening, our new bril representation also assigns each variable a number (zero-indexed per function) instead of strings, which makes it easy for us to apply bitset optimization. For block b, bit `i` in `in[b]` means either “definition at `function.instruction[i]`” reaches b (reaching definition) or “variable `i` is live at b” (liveness analysis). We used a [SIMD accelerated bitset](https://docs.rs/fixedbitset/latest/fixedbitset/) implementation for efficiency. 
+
+For both the sequential baseline and the parallel version, we only compute KILL and GEN sets once for each block before running the dataflow solver and use them afterwards in all transfer passes. This avoids re-iterating block’s instruction on every transfer whenever the `in[b]` is changed. Both the reaching definition  and liveness analysis share the same transfer function given KILL and GEN set:
+```
+transfer(b) = (in[b] \ KILL[b]) U GEN[b]
+```
+Another important observation is that: most of the computations for KILL and GEN are embarrassingly parallelizable. 
+
+
+**Reaching definition**:
+* `GEN[b]`: a set of local variables defined in block b
+* `KILL[b]`: for definition `d: y = … in b, KILL[b][d] = DEFS[y] - {d}`
+
+`GEN[b]` only depends on block local info. `KILL[b]` requires `DEFS[y]` across every block, while `DEFS[y]` can be computed with a simple map-reduce or fold-reduce (however, empirically, we found that fold-reduce/map-reduce has worse performance than the sequential baseline in our setting)
```

Comment:
I think I don't understand what `DEFS[y]` is. Is there any way you could explain farther how this works, perhaps by defining all your terms clearly?